src/common/transformations/src/transformations/smart_reshape/reshape_sinking.cpp

```diff
@@ -0,0 +1,152 @@
+            return false;
+        auto input_rank = input_pshape.size();
+        if (input_pshape[input_rank - 1] != K)
+            return false;
+
+        // check matmul eligibility: has constant second input in a form of [O, K]
+        auto matmul = dynamic_pointer_cast<MatMul>(pattern_to_node.at(matmul_label));
+        if (!matmul || matmul->get_transpose_a())
+            return false;
+        int64_t O = -1;
+        if (const auto& constant = dynamic_pointer_cast<Constant>(matmul->get_input_node_shared_ptr(1))) {
+            const auto& constant_shape = constant->get_shape();
+            if (constant_shape.size() != 2)
+                return false;
+            const auto& desired_K_index = matmul->get_transpose_b() ? 1 : 0;
+            const auto& O_index = matmul->get_transpose_b() ? 0 : 1;
+            if (constant_shape[desired_K_index] != K)
+                return false;
+            O = static_cast<int64_t>(constant_shape[O_index]);
+        }
+        if (O == -1)
+            return false;
+
+        // check add eligibility if present: has constant second input that has a form of [1, 1, ..., O] (doesn't
+        // broadcast first input)
+        if (pattern_to_node.count(add_label)) {
+            auto add = dynamic_pointer_cast<Add>(pattern_to_node.at(add_label));
+            if (!add || add->get_autob() != ov::op::AutoBroadcastType::NUMPY)
+                return false;
+            const auto& constant = dynamic_pointer_cast<Constant>(add->get_input_node_shared_ptr(1));
+            if (!constant)
+                return false;
+            const auto& constant_shape = constant->get_shape();
+            auto desired_ones_shape = ov::Shape(constant_shape.size(), 1);
+            auto desired_shape = ov::Shape(constant_shape.size() - 1, 1);
+            desired_shape.push_back(O);
+            OPENVINO_ASSERT(constant_shape.size() == desired_ones_shape.size() &&
+                            constant_shape.size() == desired_shape.size());
+            if (constant_shape != desired_shape && constant_shape != desired_ones_shape)
+                return false;
+        }
+
+        // check second Reshape eligibility: has hard-coded output pattern constant which is almost the same as
+        // input_shape of the pattern except for the batch and last dimension
+        auto reshape_1 = m.get_match_root();
+
+        const auto& constant = dynamic_pointer_cast<Constant>(reshape_1->get_input_node_shared_ptr(1));
+        if (constant == nullptr)
+            return false;
+        auto output_pattern = constant->cast_vector<int64_t>();
+        if (output_pattern.size() != input_rank)
+            return false;
+        if (!all_of(output_pattern.begin(), output_pattern.end(), [](const int64_t& i) {
+                return i > 0;
+            }))
+            return false;
+        for (size_t i = 1; i < input_rank; ++i) {
+            if (i + 1 == input_rank) {
```

Comment:
nitpick: use output_patter[input_rank] != O, not nessary to check it in the loop